### PR TITLE
parser: database.ReservedKeyword as identifier & clean up

### PR DESCRIPTION
**Pull request by tiancaiamao**
1. fix a bug test.desc can't be used as identifier, which is valid in mysql
2. introduce ReservedKeyword in parser.y and clean up

desc is a reserved keyword, before fix

```
mysql> select * from test.desc;
ERROR 1105 (HY000): line 0 column 23 near ""
```

after fix

```
mysql> select * from test.desc;
Empty set (0.00 sec)
```

### CODE REVIEW

File: parser/parser.y [hunk 1]
```diff
@@ -2005,6 +1994,31 @@ UnReservedKeyword:
...
+| "LOCALTIMESTAMP" | "LOCK" | "LONGBLOB" | "LONGTEXT" | "MEDIUMBLOB" | "MEDIUMINT" | "MEDIUMTEXT"
+| "MINUTE_MICROSECOND" | "MINUTE_SECOND" | "MOD" | "NOT" | "NO_WRITE_TO_BINLOG" | "NULL" | "NUMERIC"
+| "ON" | "OPTION" | "OR" | "ORDER" | "OUTER" | "PRECISION" | "PRIMARY" | "PROCEDURE" | "READ" | "REAL"
+| "REFERENCES" | "REGEXP" | "REPEAT" | "REPLACE" | "RESTRICT" | "RIGHT" | "RLIKE"
+| "SCHEMA" | "SCHEMAS" | "SECOND_MICROSECOND" | "SELECT" | "SET" | "SHOW" | "SMALLINT"
+| "STARTING" | "TERMINATED" | "THEN" | "TINYBLOB" | "TINYINT" | "TINYTEXT" | "TO"
+| "TRAILING" | "TRUE" | "UNION" | "UNIQUE" | "UNLOCK" | "UNSIGNED"
+| "UPDATE" | "USE" | "USING" | "UTC_DATE" | "VALUES" | "VARBINARY" | "VARCHAR"
+| "WHEN" | "WHERE" | "WRITE" | "XOR" | "YEAR_MONTH" | "ZEROFILL"
+ /*
```
Comment (shenli) [inline parser/parser.y]: Why comment this line?

File: parser/parser_test.go [hunk 2]
```diff
@@ -727,6 +727,9 @@ func (s *testParserSuite) TestIdentifier(c *C) {
 		{`select 1 a, 1 "a", 1 'a'`, true},
 		{`select * from t as "a"`, false},
 		{`select * from t a`, true},
+		// reserved keyword can't be used as identifier directly, but A.B pattern is an exception
+		{`select COUNT from DESC`, false},
+		{`select COUNT from SELECT.DESC`, true},
```
Comment (shenli) [inline parser/parser_test.go]: Is 'select desc.desc from desc.desc;' a valid sql?
db, table, column name are all desc.

File: parser/parser.y (hunk 1 above)
Comment (tiancaiamao) [inline parser/parser.y]: Those are ReservedKeyword should be add.
but goyacc would report shift/reduce error and fail...
I'll find a way to fix this.

File: parser/parser_test.go (hunk 2 above)
Comment (tiancaiamao) [inline parser/parser_test.go]: no, desc.desc can be column name...

**Comment by tiancaiamao** [comment]
it's strange that:

```
mysql> use `desc`;
ERROR 1105 (HY000): line 0 column 8 near ""
```

I'll look into it tomorrow

File: parser/parser_test.go (hunk 2 above)
Comment (shenli) [inline parser/parser_test.go]: Could you add unit test like: https://github.com/pingcap/tidb/blob/master/parser/parser_test.go#L55
So all the keywords will be covered.

**Comment by tiancaiamao** [comment]
PTAL @coocood @shenli

**Comment by shenli** [comment]
LGTM
@coocood @zimulala PTAL

**Comment by coocood** [comment]
LGTM

**Pull request merged**